### PR TITLE
OCSP: Disable when NODE_TLS_REJECT_UNAUTHORIZED is set to 0

### PR DESCRIPTION
**Pull request by orgads**
I'm not sure if this is the right way to solve it. Feel free to suggest alternatives :)

### CODE REVIEW

File: external/ocsp/ocsp/agent.js [hunk 1]
```diff
@@ -67,6 +67,10 @@ Agent.prototype.createConnection = function createConnection(port,
 Agent.prototype.handleOCSPResponse = function handleOCSPResponse(socket,
                                                                  stapling,
                                                                  cb) {
+  if (process.env.NODE_TLS_REJECT_UNAUTHORIZED === '0') {
+    cb();
+    return;
```
Comment (rhurey) [inline external/ocsp/ocsp/agent.js]: This should likely go in WebSocketMessageAdapter to not register the agent in the first place if checks are going to be skipped.

There's two reasons for this:
1) I'd like to keep as close to the ocsp library as possible in the hopes of getting the changes made so far committed back to the origin.

2) This doesn't cover the scenario where a proxy is used.

Comment (orgads) [reply external/ocsp/ocsp/agent.js]: Ok, I pushed a revised commit. Is it better?

File: src/common.browser/WebsocketMessageAdapter.ts [hunk 2]
```diff
@@ -95,6 +95,7 @@ export class WebsocketMessageAdapter {
         this.privConnectionState = ConnectionState.Connecting;
 
         try {
+            const enableOCSP = process.env.NODE_TLS_REJECT_UNAUTHORIZED !== "0";
```
Comment (rhurey) [inline src/common.browser/WebsocketMessageAdapter.ts]: >const enableOCSP [](start = 12, length = 16)

const enableOCSP :boolean

File: src/common.browser/WebsocketMessageAdapter.ts [hunk 3]
```diff
@@ -123,34 +126,37 @@ export class WebsocketMessageAdapter {
...
                 } else {
-                    // The ocsp library will handle validation for us and fail the connection if needed.
-                    this.privCertificateValidatedDeferral.resolve(true);
-
-                    const ocspAgent: ocsp.Agent = new ocsp.Agent({});
-                    const options: ws.ClientOptions = { agent: ocspAgent };
+                    const options: ws.ClientOptions = {};
+                    if (enableOCSP) {
+                        // The ocsp library will handle validation for us and fail the connection if needed.
+                        this.privCertificateValidatedDeferral.resolve(true);
```
Comment (rhurey) [inline src/common.browser/WebsocketMessageAdapter.ts]: Ran the test suite with NODE_TLS_REJECT_UNAUTHORIZED = 0 and all the tests that use the service failed. 

I suspect it's due to this deferral never being resulted if enable OCSP is false. For the non-proxy case it should always be resolved, and in the proxy case resolved if enableOCSP is false.

**Comment by rhurey** [comment]
It's been 4 months, I'm going to reject this PR if there isn't a response to comments in a week.

**Comment by orgads** [comment]
Hi, sorry for the delay. I'll try to finish it this week.

**Comment by orgads** [comment]
Next step is to make it configurable (possibly with default set according to `NODE_TLS_REJECT_UNAUTHORIZED`), but I'm not familiar with the code, so I'll leave it to you. :)

**Comment by orgads** [comment]
@rhurey: My turn to ping you ;)

**Comment by rhurey** [comment]
Closing and reopening to clear some Github->DevOps integration snags.

**Comment by rhurey** [comment]
https://msasg.visualstudio.com/Skyman/_build?definitionId=7863 Passed on merge, not sure why GitHub integration won't see it. Still need to test with env enabled.

**Pull request merged**